Implement loading custom css in the GTK app

b6k-dev opened the pull request:
Closes https://github.com/ghostty-org/ghostty/issues/4089
Gave it a shot and implemented the custom css loading.
My general idea is to use a provider for each stylesheet the user wants to load and then when the config changes unload them and create new providers.
A separate provider has to be used for each stylesheet the user wants to load, since when the provider loads the css it clears all the previously loaded styles, so in effect we cannot use one provider to load multiple stylesheets, but maybe there is a better way to overcome this limitation which I'm not seeing.

b6k-dev committed:
Implement loading custom css in the GTK app

maciekbartczak commented on src/apprt/gtk/App.zig:
```diff
@@ -1040,6 +1055,44 @@ fn loadRuntimeCss(
+            .optional => |path| .{ path, true },
+            .required => |path| .{ path, false },
+        };
+        std.fs.accessAbsolute(path, .{}) catch |err| {
+            if (err != error.FileNotFound or !optional) {
+                log.err("error opening gtk-custom-css file {s}: {}", .{ path, err });
+            }
+            continue;
+        };
```
I'm not sure if this is the place where this kind of validation should happen. I'm not that familiar with the codebase yet, but maybe this should be validated when the config is loaded, so that the message could be appended to `_diagnostics`?

jcollie replied:
There's a bit of a race condition between the access test and the usage here. I was playing around with this and well and I think that the better option is to open and read the file using Zig functions and then use `gtk_css_provider_load_from_data` or `gtk_css_provider_load_from_bytes`.

We probably should have a helper function that takes a `[]const u8` and then uses `gtk_css_provider_load_from_data` or `gtk_css_provider_load_from_bytes` depending on GTK version (one is deprecated in 4.12 and one is not available until 4.12). That can then be used in all the places that we load CSS.

jcollie replied:
> maybe this should be validated when the config is loaded, so that the message could be appended to `_diagnostics`?

Perhaps when the config is loaded just do an access check and add a diagnostic if the file isn't optional, but I think that's not necessary for this PR.

maciekbartczak replied:
Then I'll leave the scope of this PR as is and won't mess with the validation during config load.

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -81,6 +81,9 @@ transient_cgroup_base: ?[]const u8 = null,
 
+/// Providers for loading custom stylesheets defined by user
+custom_css_providers: std.ArrayList(*c.GtkCssProvider),
+
```
I'd recommend using a `std.ArrayListUnmanaged` instead.

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -425,6 +428,7 @@ pub fn init(core_app: *CoreApp, opts: Options) !App {
         .running = c.g_application_get_is_remote(gapp) == 0,
         .css_provider = css_provider,
+        .custom_css_providers = std.ArrayList(*c.GtkCssProvider).init(core_app.alloc),
     };
```
If a `std.ArrayListUnmanaged` used then it can have a default value of `.{}` and then doesn't need to be initialized here.

b6k-dev committed:
code review:
- use ArrayListUnmanaged
- read the stylesheet file using zig api
- use proper css_provider_load_ function depending on the GTK version

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -1073,23 +1066,50 @@ fn loadCustomCss(self: *App) Allocator.Error!void {
+        const contents = try buf_reader.reader().readAllAlloc(
+            self.core_app.alloc,
+            5 * 1024 * 1024 // 5MB
+        );
```
Should note in the config entry comments about the file size limit. I mean 5MiB is _probably_ enough, but it would be good to document that limitation.

maciekbartczak replied:
Added the info in the config.

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -82,7 +82,7 @@ transient_cgroup_base: ?[]const u8 = null,
 /// Providers for loading custom stylesheets defined by user
-custom_css_providers: std.ArrayList(*c.GtkCssProvider),
+custom_css_providers: std.ArrayListUnmanaged(*c.GtkCssProvider),
 
```
You can give this a default value as well if you want.

maciekbartczak replied:
Sure, initializing it here with a default value is fine.

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -428,7 +428,7 @@ pub fn init(core_app: *CoreApp, opts: Options) !App {
         .css_provider = css_provider,
-        .custom_css_providers = std.ArrayList(*c.GtkCssProvider).init(core_app.alloc),
+        .custom_css_providers = .{},
     };
```
Not necessary if you add a default value above, but it's more of a nitpick on my part.

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -1073,23 +1066,50 @@ fn loadCustomCss(self: *App) Allocator.Error!void {
-        const provider = c.gtk_css_provider_new();
+        log.info("loading gtk-custom-css path={s}", .{path});
+        var buf_reader = std.io.bufferedReader(file.reader());
+        const contents = try buf_reader.reader().readAllAlloc(
```
I think that the buffered reader is unnecessary.

maciekbartczak replied:
switched to a non-buffered reader

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -909,11 +909,8 @@ fn syncConfigChanges(self: *App) !void {
-            .{},
-        ),
+    self.loadCustomCss() catch |err| {
+        log.warn("Failed to load custom CSS, no custom CSS applied, err={}", .{err});
```
We are trying to be more explicit about catching exceptions, so unless something added dozens of errors we should keep the switch. It helps catch regressions if something accidentally starts returning more errors.

maciekbartczak replied:
I implemented your suggestion, however i have some doubts - maybe there is a simpler way to handle the Reader errors. At first I thought handling all errors from an error set would be possible with one case, something like:
```
    self.loadCustomCss() catch |err| switch (err) {
        std.fs.File.Reader.Error => log.warn(
            "failed to load custom CSS, no custom CSS applied - encountered error while reading file: {}",
            .{err},
        ),
    };
```
I tried a few different ways to handle all the errors in one case, however the compiler complained and forced me to handle each error from the set separately, so i just did a one big case:
```
        std.fs.File.Reader.Error.InputOutput,
        std.fs.File.Reader.Error.SystemResources,
        std.fs.File.Reader.Error.IsDir,
        std.fs.File.Reader.Error.OperationAborted,
        std.fs.File.Reader.Error.BrokenPipe,
        std.fs.File.Reader.Error.ConnectionResetByPeer,
        std.fs.File.Reader.Error.ConnectionTimedOut,
        std.fs.File.Reader.Error.NotOpenForReading,
        std.fs.File.Reader.Error.SocketNotConnected,
        std.fs.File.Reader.Error.WouldBlock,
        std.fs.File.Reader.Error.AccessDenied => log.warn(
            "failed to load custom CSS, no custom CSS applied - encountered error while reading file: {}",
            .{err},
        ),
```
If this is not a standard way to handle error sets in switch statements please point me to the correct solution :smile:

jcollie replied:
Yeah, I didn't think that it would get that out of hand. Feel free to revert this part.

maciekbartczak replied:
Reverted.

b6k-dev committed:
code review:
- initialize custom_css_providers using a default value
- remove usage of buffered reader
- document maximum file size
- handle exceptions explicitly

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -909,8 +908,33 @@ fn syncConfigChanges(self: *App) !void {
+        std.fs.File.Reader.Error.SocketNotConnected,
+        std.fs.File.Reader.Error.WouldBlock,
+        std.fs.File.Reader.Error.AccessDenied => log.warn(
+            "failed to load custom CSS, no custom CSS applied - encountered error while reading file: {}",
```
OK, that's maybe a case that you don't need the switch, sorry!

b6k-dev committed:
code review - revert explicit error handling

jcollie commented on src/config/Config.zig:
```diff
@@ -1925,6 +1925,14 @@ keybind: Keybinds = .{},
+/// Prepend a ? character to the file path to suppress errors if the file does
+/// not exist. If you want to include a file that begins with a literal ?
+/// character, surround the file path in double quotes (").
+@"gtk-custom-css": RepeatablePath = .{},
```
I was doing some testing and I had a crash if I didn't explicitly specify an absolute path to the custom CSS file, so for some reason the paths are not being expanded properly.

rhodes-b replied:
https://github.com/ghostty-org/ghostty/pull/3811 should fix expansion not sure about the crash

jcollie replied:
> #3811 should fix expansion not sure about the crash

The crash is specifically because we have an assert in place to check for non-absolute paths. And #3811 won't fix it because I wasn't using a `~`, just a plain filename `--gtk-custom-css=custom.css`. `RepeatablePath` _should_ resolve that relative to Ghostty's config dir I thought but it obviously didn't.

jcollie replied:
Actually, I think that I figured out my problem, if you specify `--background-image=custom.css` on the CLI, Ghostty tries to resolve that based on the `$PWD`. If it can't be resolved, I think that the code leaves it as is without any warnings and then we eventually get a panic. So that's not a blocker for this (even though Mitchell has already merged this).
